nova_plugin/server.py

```diff
@@ -548,11 +548,21 @@ def attach_volume(nova_client, cinder_client, **kwargs):
         server_id,
         volume_id,
         device if device != 'auto' else None)
-    volume.wait_until_status(cinder_client=cinder_client,
-                             volume_id=volume_id,
-                             status=volume.VOLUME_STATUS_IN_USE)
-    if device == 'auto':
+    try:
+        vol, wait_succeeded = volume.wait_until_status(
+            cinder_client=cinder_client,
+            volume_id=volume_id,
+            status=volume.VOLUME_STATUS_IN_USE
+        )
+        if not wait_succeeded:
+            raise RecoverableError(
+                'Waiting for volume status {0} failed - detaching volume and '
+                'retrying..'.format(volume.VOLUME_STATUS_IN_USE))
+    except RecoverableError:
```

Comment:
The explanation for the 2nd point was sure clear enough ;) I kept the fix locally but now I have pushed it (+ test).

Now I will think about how to deal with the first point. You are probably right that we should convert any exception thrown in cleanup to a NonRecoverable one..